pkg/scd/operational_intents_handler.go

```diff
@@ -506,53 +549,60 @@ func (a *Server) upsertOperationalIntentReference(ctx context.Context, authorize
 					subToUpsert.NotifyForConstraints = *params.NewSubscription.NotifyForConstraints
 				}
 
-				sub, err = r.UpsertSubscription(ctx, &subToUpsert)
+				effectiveSubscription, err = r.UpsertSubscription(ctx, &subToUpsert)
 				if err != nil {
 					return stacktrace.Propagate(err, "Failed to create implicit subscription")
 				}
 			}
 
 		} else {
-			// Use existing Subscription
-			sub, err = r.GetSubscription(ctx, subscriptionID)
-			if err != nil {
-				return stacktrace.Propagate(err, "Unable to get Subscription")
-			}
-			if sub == nil {
-				return stacktrace.NewErrorWithCode(dsserr.BadRequest, "Specified Subscription %s does not exist", subscriptionID)
+			// A subscription has been specified:
+			// If it is different from the previous subscription, we need to fetch it from the store
+			// in order to ensure it correctly covers the OIR.
+			if effectiveSubscription == nil || previousSubIsBeingReplaced {
```

Comment:
Actually now that I look at it again, I'm also not completely sure why this change here is needed? Was there previously an issue that is fixed by this change? Or is it just an optimization to avoid fetching twice the same subscription?

Reply:
> effectiveSubscription == nil will always be true here.

Thanks for catching. A previous revision was pre-setting the value and that ended up being even more confusing than it is now.

> Actually now that I look at it again, I'm also not completely sure why this change here is needed? Was there previously an issue that is fixed by this change? Or is it just an optimization to avoid fetching twice the same subscription?

This is to avoid re-fetching the same subscription already fetched at line 496 in the `if old.SubscriptionID != nil {` block if that subscription is not being replaced. In cases where it is being replaced, we need to fetch the subscription that will eventually be attached to the OIR to verify that it properly covers the OIR.

It indeed ensures that we don't do unnecessary roundtrips.

Reply:
> Thanks for catching. A previous revision was pre-setting the value and that ended up being even more confusing than it is now.

(Well, as a good illustration of how convoluted the control logic got, my 'simplification' was too much and needs a rework)

Reply:
> This is to avoid re-fetching the same subscription already fetched at line 496 in the if old.SubscriptionID != nil { block if that subscription is not being replaced. In cases where it is being replaced, we need to fetch the subscription that will eventually be attached to the OIR to verify that it properly covers the OIR.
> 
> It indeed ensures that we don't do unnecessary roundtrips.

Could you add a comment mentioning that? That clarifies that this is not actually a separate case, but an optimization. Ideally that could be hidden behind a function but not sure if that's doable or makes sense here.